blockchain/indexers/utreexocfindex.go

```diff
@@ -0,0 +1,379 @@
+// filter index.
+func (idx *UtreexoCFIndex) Create(dbTx database.Tx) error {
+	meta := dbTx.Metadata()
+
+	utreexoCfIndexParentBucket, err := meta.CreateBucket(utreexoCFIndexParentBucketKey)
+	if err != nil {
+		return err
+	}
+
+	for _, bucketName := range utreexoCfHeaderKeys {
+		_, err = utreexoCfIndexParentBucket.CreateBucket(bucketName)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// storeUtreexoCFilter stores a given utreexocfilter header
+func storeUtreexoCFHeader(dbTx database.Tx, block *btcutil.Block, filterData []byte,
+	filterType wire.FilterType) error {
+	if filterType != wire.UtreexoCFilter {
+		return errors.New("invalid filter type")
+	}
+
+	// Figure out which header bucket to use.
+	hkey := utreexoCfHeaderKeys[0]
+	h := block.Hash()
+
+	// fetch the previous block's filter header.
+	var prevHeader *chainhash.Hash
+	ph := &block.MsgBlock().Header.PrevBlock
+	if ph.IsEqual(&zeroHash) {
+		prevHeader = &zeroHash
+	} else {
+		pfh, err := dbFetchUtreexoCFilterIdxEntry(dbTx, hkey, ph)
+		if err != nil {
+			return err
+		}
+
+		// Construct the new block's filter header, and store it.
+		prevHeader, err = chainhash.NewHash(pfh)
+		if err != nil {
+			return err
+		}
+	}
+
+	fh, err := builder.MakeHeaderForUtreexoCFilter(filterData, *prevHeader)
+	if err != nil {
+		return err
+	}
+	return dbStoreUtreexoCFilterIdxEntry(dbTx, hkey, h, fh[:])
+}
+
+// ConnectBlock is invoked by the index manager when a new block has been
+// connected to the main chain.
+// This is part of the Indexer interface.
+func (idx *UtreexoCFIndex) ConnectBlock(dbTx database.Tx, block *btcutil.Block,
+	stxos []blockchain.SpentTxOut) error {
+
+	blockHash := block.Hash()
+	roots, leaves, err := idx.fetchUtreexoRoots(dbTx, blockHash)
+
+	if err != nil {
+		return err
+	}
+
+	// serialize the hashes of the utreexo roots hash
+	serializedUtreexo, err := blockchain.SerializeUtreexoRootsHash(leaves, roots)
+	if err != nil {
+		return err
+	}
+
+	return storeUtreexoCFHeader(dbTx, block, serializedUtreexo, wire.UtreexoCFilter)
+}
+
+// fetches the utreexo roots for a given block hash
+func (idx *UtreexoCFIndex) fetchUtreexoRoots(dbTx database.Tx,
+	blockHash *chainhash.Hash) ([]*chainhash.Hash, uint64, error) {
+
+	var leaves uint64
+	var roots []*chainhash.Hash
+
+	// For compact state nodes
+	if !idx.noUtreexo {
+		viewPoint, err := idx.chain.FetchUtreexoViewpoint(blockHash)
+		if err != nil {
+			return nil, 0, err
+		}
+		roots = viewPoint.GetRoots()
+		leaves = viewPoint.NumLeaves()
+		return roots, leaves, nil
+	}
+	// for bridge nodes
+	if idx.utreexoProofIndex != nil {
+		// log.Infof("Is utreexoProofIndex JABURAT %v", blockHash)
+		roots, leaves, err := idx.utreexoProofIndex.FetchUtreexoState(dbTx, blockHash)
+		if err != nil {
+			return nil, 0, err
+		}
+		return roots, leaves, nil
+	} else if idx.flatUtreexoProofIndex != nil {
+		height, err := idx.chain.BlockHeightByHash(blockHash)
```

Comment:
Ok so the error happens here but I don't think this code is fetching the correct utreexo roots because of this weird quirk: https://github.com/utreexo/utreexod/blob/f68f79df9d49a7d7cd329b86107767c3882d2b88/rpcserver.go#L3502-L3511

Notice how the roots are being fetched:
https://github.com/utreexo/utreexod/blob/f68f79df9d49a7d7cd329b86107767c3882d2b88/rpcserver.go#L3512-L3513
https://github.com/utreexo/utreexod/blob/f68f79df9d49a7d7cd329b86107767c3882d2b88/rpcserver.go#L3529-L3530

# The actual bug and the fix

This code is causing the error because the node hasn't been included in `bestChain` yet. That happens after the indexes are connected https://github.com/utreexo/utreexod/blob/f68f79df9d49a7d7cd329b86107767c3882d2b88/blockchain/chain.go#L777-L778

You have access to height so `block.Height()` shouldn't result in an error but I think you should really be calling `FetchCurrentUtreexoState()`

Reply:
I guess I'm not quite sure if this is how the roots should be fetched in the first place because I think it's weird that we're fetching from the utreexo state here and the indexers interface doesn't have any guarantees that the utreexocfindex gets called after flatutreexoproofindex or the utreexoproofindex.

Reply:
> I guess I'm not quite sure if this is how the roots should be fetched in the first place because I think it's weird that we're fetching from the utreexo state here and the indexers interface doesn't have any guarantees that the utreexocfindex gets called after flatutreexoproofindex or the utreexoproofindex.

The indexers are appended to the indexes slice right, and connecting the blocks is in the order with which those indexes are outlined in the slice. Doesn't that mean there's a particular order for which those indexes get called?